RFC: [ADT] Use a storage policy in DenseMap/SmallDenseMap (NFC)

kazutakahirata opened the pull request:
This patch refactors DenseMap and SmallDenseMap to a policy-based
design, moving away from the previous CRTP-based implementation.

The primary motivation is to improve the architecture by centralizing
logic and removing code duplication.  In the current implementation,
DenseMap and SmallDenseMap use CRTP, which splits responsibilities in
a way that forces redundant code.  Specifically, all lifetime
management logic (constructors, destructor, copy/move assignment
operators) is implemented separately in both DenseMap and
SmallDenseMap, while the core container logic resides in DenseMapBase.

This patch resolves this by introducing a clean separation of concerns
using a storage policy design:

- DenseMapBase: Is now a generic, policy-driven implementation that
  contains all user-facing logic, including constructors and lifetime
  management.  It is parameterized by a storage policy.

- DenseMapStorage / SmallDenseMapStorage: Are dumb storage policies
  responsible only for memory management (allocation/deallocation) and
  metadata storage. They are owned by DenseMapBase via composition.

- DenseMap / SmallDenseMap: Become simple, user-facing forwarding
  classes (akin to type aliases) that instantiate DenseMapBase with
  the appropriate storage policy.

Key Benefits of this Refactoring:

- Improved Code Reuse: All constructors, the destructor, and
  assignment operators are now implemented once in DenseMapBase,
  eliminating code duplication.

- Clear Separation of Concerns: The logic for hashing and iteration is
  now cleanly separated from memory allocation and storage layout.

- Reduced Code Size: This refactoring reduces the line count of
  DenseMap.h by 137 lines (~10%).

Comments would be greatly appreciated. (Of course, I'm not planning to
post this as a single giant pull request.)

tgymnich reviewed (commented):
Good improvement. Thank you!

kuhar reviewed (commented):
Makes sense

kazutakahirata commented on llvm/include/llvm/ADT/DenseMap.h:
```diff
@@ -350,7 +392,8 @@ class DenseMapBase : public DebugEpochBase {
   /// value in the DenseMap).
   [[nodiscard]] bool isPointerIntoBucketsArray(const void *Ptr) const {
-    return Ptr >= getBuckets() && Ptr < getBucketsEnd();
+    auto Buckets = buckets();
```
I'll incorporate this when I post split-up patches.  Thanks!

s-barannikov commented on llvm/include/llvm/ADT/DenseMap.h:
```diff
@@ -518,43 +566,46 @@ class DenseMapBase : public DebugEpochBase {
 
-  void grow(unsigned AtLeast) { derived().grow(AtLeast); }
+  void grow(unsigned AtLeast) {
+    AtLeast = Storage.computeNumBuckets(AtLeast);
```
```suggestion
    AtLeast = StorageT::computeNumBuckets(AtLeast);
```

kazutakahirata committed:
RFC: [ADT] Use a storage policy in DenseMap/SmallDenseMap (NFC)

This patch refactors DenseMap and SmallDenseMap to a policy-based
design, moving away from the previous CRTP-based implementation.

The primary motivation is to improve the architecture by centralizing
logic and removing code duplication.  In the current implementation,
DenseMap and SmallDenseMap use CRTP, which splits responsibilities in
a way that forces redundant code.  Specifically, all lifetime
management logic (constructors, destructor, copy/move assignment
operators) is implemented separately in both DenseMap and
SmallDenseMap, while the core container logic resides in DenseMapBase.

This patch resolves this by introducing a clean separation of concerns
using a storage policy design:

- DenseMapBase: Is now a generic, policy-driven implementation that
  contains all user-facing logic, including constructors and lifetime
  management.  It is parameterized by a storage policy.

- DenseMapStorage / SmallDenseMapStorage: Are dumb storage policies
  responsible only for memory management (allocation/deallocation) and
  metadata storage. They are owned by DenseMapBase via composition.

- DenseMap / SmallDenseMap: Become simple, user-facing forwarding
  classes (akin to type aliases) that instantiate DenseMapBase with
  the appropriate storage policy.

Key Benefits of this Refactoring:

- Improved Code Reuse: All constructors, the destructor, and
  assignment operators are now implemented once in DenseMapBase,
  eliminating code duplication.

- Clear Separation of Concerns: The logic for hashing and iteration is
  now cleanly separated from memory allocation and storage layout.

- Reduced Code Size: This refactoring reduces the line count of
  DenseMap.h by 137 lines (~10%).

Comments would be greatly appreciated. (Of course, I'm not planning to
post this as a single giant pull request.)
